### ISSUE TITLE
Support of VS2017-64bit (getting rid of warnings)

### ISSUE THREAD

**Issue by klotzt-draeger** [openssl/openssl#8236]
Current master issues some warnings (mostly about size_t and int not being the same) when compiling with Visual Studio 2017 (Microsoft (R) C/C++ Optimizing Compiler Version 19.16.27026.1 for x64)
Either the compiler flags (not prefered) or the code would need to be adapted to be able to compile (prefered).

Examples for compile issues are (one out of many many)
```
crypto\asn1\a_object.c(65): error C2220: warning treated as error - no 'object' file generated
crypto\asn1\a_object.c(65): warning C4267: '=': conversion from 'size_t' to 'int', possible loss of data
```

**Comment by klotzt-draeger** [comment]
one of the major problems seems to be the unclean usage of size_t vs. int (or long or unsigned int) in some files.
woudl a way forward be to make it more consistent and use size_t everywhere?

**Comment by mattcaswell** [comment]
> woudl a way forward be to make it more consistent and use size_t everywhere?

Yes, but unfortunately that is easier said than done. We did do a major effort for the 1.1.1 release to make libssl size_t clean throughout. This was a major job. But libcrypto was not tackled because it is even bigger. Part of the problem is that there is inconsistent use of different integer types, int, long, size_t, unsigned int etc baked into the API (which we can't really change without breaking user apps). 

I'd love to do libcrypto too, but so far it's not made it to the top of the priority list.

**Comment by klotzt-draeger** [comment]
agreed, this is all but a small thing. it also is onyl an issue for the 64bit version, therefore the need might not be as pressing...

**Comment by AshleyT3** [comment]
Saw these myself, fwiw a distillation...

### Summary by warning (duplicates for file/line removed)

|Count|Warning|Description|
|---:|---|---|
|958| warning C4267|conversion, possible loss of data|
|92| warning C4244|conversion, possible loss of data|
|6| warning C4133|incompatible types (usually one volatile pointer type to another)|
|1| warning C4334|32-bit shift implicitly converted to 64 bits (intended?)|
|1| warning C4311|pointer truncation (the one case impies void** is safely int*)|

### Summary by source area (duplicates for file/line removed)

| Count | Area |
| ---: | --- |
|438|crypto|
|278|test|
|163|providers|
|68|ssl|
|49|fuzz|
|41|apps|
|19|engines|
|1|include|
|1|ms|

Source data: [nmake_output_warnings_sorted_unique.txt](https://github.com/openssl/openssl/files/14033545/nmake_output_warnings_sorted_unique.txt)

**Comment by t8m** [comment]
This was fixed on 3.6.
